File: solutions/typescript/matrix/1/matrix.ts
export class Matrix {
  private matrix: number[][];

  constructor(matrixString: string) {
    // Split the string into rows by newline
    const rows = matrixString.split('\n');
    
    // Parse each row into an array of numbers
    this.matrix = rows.map(row => 
      row.trim().split(/\s+/).map(Number)
    );
  }

  get rows(): number[][] {
    // Return a copy of the rows to prevent external modification
    return this.matrix.map(row => [...row]);
  }

  get columns(): number[][] {
    const numRows = this.matrix.length;
    const numCols = this.matrix[0]?.length || 0;
    const columns: number[][] = [];
    
    // Build each column by collecting the same index from each row
    for (let col = 0; col < numCols; col++) {
      const column: number[] = [];
      for (let row = 0; row < numRows; row++) {
        column.push(this.matrix[row][col]);
      }
      columns.push(column);
    }
    
    return columns;
  }
}

File: solutions/java/matrix/1/src/main/java/Matrix.java
class Matrix {
    private final int[][] matrix;
    
    Matrix(String matrixAsString) {
        // Handle empty input
        if (matrixAsString == null || matrixAsString.isEmpty()) {
            this.matrix = new int[0][0];
            return;
        }
        
        // Split by newline to get rows
        String[] rows = matrixAsString.split("\n");
        this.matrix = new int[rows.length][];
        
        // Parse each row into integers
        for (int i = 0; i < rows.length; i++) {
            // Split by one or more whitespace characters
            String[] values = rows[i].trim().split("\\s+");
            matrix[i] = new int[values.length];
            for (int j = 0; j < values.length; j++) {
                matrix[i][j] = Integer.parseInt(values[j]);
            }
        }
    }

    int[] getRow(int rowNumber) {
        // Return a defensive copy (1-indexed input)
        return matrix[rowNumber - 1].clone();
    }

    int[] getColumn(int columnNumber) {
        // Extract column values (1-indexed input)
        int[] column = new int[matrix.length];
        for (int i = 0; i < matrix.length; i++) {
            column[i] = matrix[i][columnNumber - 1];
        }
        return column;
    }
}